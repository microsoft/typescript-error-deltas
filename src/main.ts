import ge = require("./getErrors");
import pu = require("./packageUtils");
import git = require("./gitUtils");
import { execAsync } from "./execUtils";
import type { GitResult, UserResult } from "./gitUtils";
import ip = require("./installPackages");
import ut = require("./userTestUtils");
import cp = require("child_process");
import fs = require("fs");
import path = require("path");

interface Params {
    /** True to post the result to Github, false to print to console.  */
    postResult: boolean;
    /**
     * Store test repos on a tmpfs.
     * Basically, the only reason not to do this would be lack of `sudo`.
     */
    tmpfs: boolean;
    /**
     * Number of repos to test, undefined for the default.
     * Git repos are chosen from Typescript-language repos based on number of stars; default is 100.
     * User repos start at the top of the list; default is all of them.
     */
    repoCount?: number | undefined;
    /**
     * The index to start counting repositories; defaults to `0`.
     * If `repoStartIndex` is 100 and `repoCount` is 100, the 100th to the 199th repos will be tested.
     */
    repoStartIndex?: number | undefined;
}
export interface GitParams extends Params {
    testType: 'git';
    oldTscVersion: string;
    newTscVersion: string;
}
export interface UserParams extends Params {
    testType: 'user';
    oldTypescriptRepoUrl: string;
    oldHeadRef: string;
    sourceIssue: number;
    requestingUser: string;
    statusComment: number;
    topRepos: boolean;
}

const skipRepos = [
    "https://github.com/storybookjs/storybook", // Too big to fit on VM
    "https://github.com/microsoft/frontend-bootcamp", // Can't be built twice in a row
    "https://github.com/BabylonJS/Babylon.js", // Runs out of space during compile
];
const processCwd = process.cwd();
const processPid = process.pid;
const executionTimeout = 10 * 60 * 1000;

// Exported for testing
export async function processRepo(params: GitParams | UserParams, topGithubRepos: boolean, downloadDir: string, userTestDir: string, repo: git.Repo, oldTscPath: string, newTscPath: string, outputs: string[]) {
    const { testType } = params
    if (params.tmpfs)
        await execAsync(processCwd, "sudo mount -t tmpfs -o size=4g tmpfs " + downloadDir);

    try {
        if (repo.url) {
            try {
                console.log("Cloning if absent");
                await git.cloneRepoIfNecessary(downloadDir, repo);
            }
            catch (err) {
                reportError(err, "Error cloning " + repo.url);
                return;
            }
        }
        else {
            await ut.copyUserRepo(downloadDir, userTestDir, repo);
        }

        const repoDir = path.join(downloadDir, repo.name);

        try {
            console.log("Installing packages if absent");
            await installPackages(repoDir, /*recursiveSearch*/ topGithubRepos, executionTimeout, repo.types);
        }
        catch (err) {
            reportError(err, `Error installing packages for ${err.packageRoot ?? "some package.json file"} in ${repo.name}`);
            await reportResourceUsage(downloadDir);
            return;
        }

        try {
            console.log(`Building with ${oldTscPath} (old)`);
            const oldErrors = await buildAndGetErrors(repoDir, oldTscPath, /*skipLibCheck*/ true, topGithubRepos, /*realTimeout*/params.postResult);

            if (oldErrors.hasConfigFailure) {
                console.log("Unable to build project graph");
                console.log(`Skipping build with ${newTscPath} (new)`);
                return;
            }

            const numProjects = oldErrors.projectErrors.length;

            let numFailed = 0;
            for (const oldProjectErrors of oldErrors.projectErrors) {
                if (oldProjectErrors.hasBuildFailure || oldProjectErrors.errors.length) {
                    numFailed++;
                }
            }

            // User tests ignores build failures.
            if (testType === "git" && numFailed === numProjects) {
                console.log(`Skipping build with ${newTscPath} (new)`);
                return;
            }

            let sawNewRepoErrors = false;
            const owner = repo.owner ? `${repo.owner}/` : "";
            const url = repo.url ? `(${repo.url})` : "";

            let repoSummary = `# [${owner}${repo.name}]${url}\n`;

            if (numFailed > 0) {
                const oldFailuresMessage = `${numFailed} of ${numProjects} projects failed to build with the old tsc`;
                console.log(oldFailuresMessage);
                repoSummary += `**${oldFailuresMessage}**\n`;
            }

            console.log(`Building with ${newTscPath} (new)`);
            const newErrors = await buildAndGetErrors(repoDir, newTscPath, /*skipLibCheck*/ true, topGithubRepos, /*realTimeout*/params.postResult);

            if (newErrors.hasConfigFailure) {
                console.log("Unable to build project graph");

                repoSummary += ":exclamation::exclamation: **Unable to build the project graph with the new tsc** :exclamation::exclamation:\n";

                outputs.push(repoSummary)
                return true;
            }

            console.log("Comparing errors");
            for (const oldProjectErrors of oldErrors.projectErrors) {
                // To keep things simple, we'll focus on projects that used to build cleanly
                if (testType === "git" && (oldProjectErrors.hasBuildFailure || oldProjectErrors.errors.length)) {
                    continue;
                }

                // TS 5055 generally indicates that the project can't be built twice in a row without cleaning in between.
                // Filter out errors reported already on "old".
                const newProjectErrors = newErrors.projectErrors.find(pe => pe.projectUrl == oldProjectErrors.projectUrl)?.errors?.filter(e => e.code !== 5055)
                    .filter(ne => !oldProjectErrors.errors.find(oe => ge.errorEquals(oe, ne)));
                if (!newProjectErrors?.length) {
                    continue;
                }

                sawNewRepoErrors = true;

                const errorMessageMap = new Map<string, ge.Error[]>();
                const errorMessages: string[] = [];

                console.log(`New errors for ${oldProjectErrors.isComposite ? "composite" : "non-composite"} project ${oldProjectErrors.projectUrl}`);
                for (const newError of newProjectErrors) {
                    const newErrorText = newError.text;

                    console.log(`\tTS${newError.code} at ${newError.fileUrl ?? "project scope"}${oldProjectErrors.isComposite ? ` in ${newError.projectUrl}` : ``}`);
                    console.log(`\t\t${newErrorText}`);

                    if (!errorMessageMap.has(newErrorText)) {
                        errorMessageMap.set(newErrorText, []);
                        errorMessages.push(newErrorText);
                    }

                    errorMessageMap.get(newErrorText)!.push(newError);
                }

                repoSummary += `### ${makeMarkdownLink(oldProjectErrors.projectUrl)}\n`
                for (const errorMessage of errorMessages) {
                    repoSummary += ` - \`${errorMessage}\`\n`;

                    for (const error of errorMessageMap.get(errorMessage)!) {
                        repoSummary += `   - ${error.fileUrl ? makeMarkdownLink(error.fileUrl) : "Project Scope"}${oldProjectErrors.isComposite ? ` in ${makeMarkdownLink(error.projectUrl)}` : ``}\n`;
                    }
                }
            }

            if (sawNewRepoErrors) {
                // sawNewErrors = true;
                // summary += repoSummary;
                outputs.push(repoSummary)
                return true;
            }
        }
        catch (err) {
            reportError(err, "Error building " + repo.url ?? repo.name);
            return;
        }

        console.log("Done " + repo.url ?? repo.name);
    }
    finally {
        // Throw away the repo so we don't run out of space
        // Note that we specifically don't recover and attempt another repo if this fails
        console.log("Cleaning up repo");
        if (params.tmpfs) {
            // Dump any processes holding onto the download directory in case umount fails
            await execAsync(processCwd, `lsof | grep ${downloadDir} || true`);
            await execAsync(processCwd, "sudo umount " + downloadDir);
            await reportResourceUsage(downloadDir);
        }
    }
}

export async function mainAsync(params: GitParams | UserParams): Promise<GitResult | UserResult | undefined> {
    const { testType } = params;

    const downloadDir = params.tmpfs ? "/mnt/ts_downloads" : "./ts_downloads";
    // TODO: check first whether the directory exists and skip downloading if possible
    // TODO: Seems like this should come after the typescript download
    if (params.tmpfs)
        await execAsync(processCwd, "sudo mkdir " + downloadDir);
    else
        await execAsync(processCwd, "mkdir " + downloadDir);

    // TODO: Only download if the commit has changed (need to map refs to commits and then download to typescript-COMMIT instead)
    const { oldTscPath, oldTscResolvedVersion, newTscPath, newTscResolvedVersion } = await downloadTypeScriptAsync(processCwd, params);

    // Get the name of the typescript folder.
    const oldTscDirPath = path.resolve(oldTscPath, "../../");
    const newTscDirPath = path.resolve(newTscPath, "../../");

    console.log("Old version = " + oldTscResolvedVersion);
    console.log("New version = " + newTscResolvedVersion);

    const userTestDir = path.join(processCwd, "userTests");

    const topGithubRepos = testType === "git" || params.topRepos;
    const repos = topGithubRepos ? await git.getPopularTypeScriptRepos(params.repoCount, params.repoStartIndex)
        : testType === "user" ? ut.getUserTestsRepos(userTestDir)
        : undefined;

    if (!repos) {
        throw new Error(`Parameter <test_type> with value ${testType} is not existent.`);
    }

    const outputs: string[] = [];
    // let summary = "";
    let sawNewErrors: true | undefined = undefined;

    let i = 0;
    const startIndex = params.repoStartIndex ?? 0;
    const maxCount = Math.min(typeof params.repoCount === 'number' ? params.repoCount : Infinity, repos.length) + startIndex;

    for (const repo of repos) {
        if (repo.url && skipRepos.includes(repo.url)) continue;
        i++;
        if (i > maxCount) break;
        console.log(`Starting #${i + startIndex} / ${maxCount}: ${repo.url ?? repo.name}`);
        if (await processRepo(params, topGithubRepos, downloadDir, userTestDir, repo, oldTscPath, newTscPath, outputs)) {
            sawNewErrors = true;
        }
    }
    const summary = outputs.join("")

    if (params.tmpfs) {
        await execAsync(processCwd, "sudo rm -rf " + downloadDir);
        await execAsync(processCwd, "sudo rm -rf " + oldTscDirPath);
        await execAsync(processCwd, "sudo rm -rf " + newTscDirPath);
    }
    else {
        await execAsync(processCwd, "rm -rf " + downloadDir);
        await execAsync(processCwd, "rm -rf " + oldTscDirPath);
        await execAsync(processCwd, "rm -rf " + newTscDirPath);
    }

    if (testType === "git") {
        const title = `[NewErrors] ${newTscResolvedVersion} vs ${oldTscResolvedVersion}`;
        const body = `The following errors were reported by ${newTscResolvedVersion}, but not by ${oldTscResolvedVersion}
[Pipeline that generated this bug](https://typescript.visualstudio.com/TypeScript/_build?definitionId=48)
[File that generated the pipeline](https://github.com/microsoft/typescript-error-deltas/blob/main/azure-pipelines-gitTests.yml)

${summary}`;
        return git.createIssue(params.postResult, title, body, !!sawNewErrors);
    }
    else if (testType === "user") {
        const body = summary
            ? `@${params.requestingUser}\nThe results of the user tests run you requested are in!\n<details><summary> Here they are:</summary><p>\n<b>Comparison Report - ${oldTscResolvedVersion}..${newTscResolvedVersion}</b>\n\n${summary}</p></details>`
            : `@${params.requestingUser}\nGreat news! no new errors were found between ${oldTscResolvedVersion}..${newTscResolvedVersion}`;
        return git.createComment(params.sourceIssue, params.statusComment, params.postResult, body);
    }
    else {
        throw new Error(`testType "${(params as any).testType}" is not a recognised test type.`);
    }
}

async function buildAndGetErrors(repoDir: string, tscPath: string, skipLibCheck: boolean, topGithubRepos: boolean, realTimeout: boolean): Promise<ge.RepoErrors> {
    if (realTimeout) {
        const p = new Promise<ge.RepoErrors>((resolve, reject) => {
            const p = cp.fork(path.join(__dirname, "run-build.js"));
            p.on('message', (m: 'ready' | ge.RepoErrors) =>
                m === 'ready'
                    ? p.send({ repoDir, tscPath, topGithubRepos, skipLibCheck })
                    : resolve(m));
            p.on('exit', reject);
        });
        return withTimeout(executionTimeout, p);
    }
    else {
        return ge.buildAndGetErrors(repoDir, tscPath, topGithubRepos, skipLibCheck)
    }
}

async function installPackages(repoDir: string, recursiveSearch: boolean, timeoutMs: number, types?: string[]) {
    let usedYarn = false;
    let timedOut = false;
    try {
        const startMs = performance.now();
        const commands = await ip.restorePackages(repoDir, /*ignoreScripts*/ true, recursiveSearch, /*lernaPackages*/ undefined, types);
        for (const { directory: packageRoot, tool, arguments: args } of commands) {
            usedYarn = usedYarn || tool === ip.InstallTool.Yarn;

            const elapsedMs = performance.now() - startMs;
            const packageRootDescription = packageRoot.substring(repoDir.length + 1) || "root directory";

            await new Promise<void>((resolve, reject) => {
                const timeout = setTimeout(async () => {
                    timedOut = true;
                    await execAsync(processCwd, `./scripts/kill-children-of ${processPid} ${tool}`);

                    const err = new Error(`Timed out after ${timeoutMs} ms`);
                    (err as any).packageRoot = packageRootDescription;
                    reject(err);
                }, timeoutMs - elapsedMs);

                cp.execFile(tool, args, { cwd: packageRoot }, err => {
                    if (!timedOut) {
                        clearTimeout(timeout);
                        if (err) {
                            (err as any).packageRoot = packageRootDescription;
                            reject(err);
                        }
                        resolve();
                    }
                });
            });
        }
    }
    finally {
        if (usedYarn) {
            await execAsync(repoDir, "yarn cache clean --all");
        }
    }
}

function withTimeout<T>(ms: number, promise: Promise<T>): Promise<T> {
    let timeout: NodeJS.Timeout | undefined;
    return Promise.race([
        promise.finally(() => timeout && clearTimeout(timeout)),
        new Promise<T>((_resolve, reject) =>
            timeout = setTimeout(async () => {
                await execAsync(processCwd, `./scripts/kill-children-of ${processPid} node`);
                return reject(new Error(`Timed out after ${ms} ms`));
            }, ms)),
    ]);
}

async function reportResourceUsage(downloadDir: string) {
    try {
        console.log("Memory");
        await execAsync(processCwd, "free -h");
        console.log("Disk");
        await execAsync(processCwd, "df -h");
        await execAsync(processCwd, "df -i");
        console.log("Download Directory");
        await execAsync(processCwd, "ls -lh " + downloadDir);
        console.log("Home Directory");
        await execAsync(processCwd, "du -csh ~/.[^.]*");
        await execAsync(processCwd, "du -csh ~/.cache/*");
    }
    catch { } // noop
}

export function reportError(err: any, message: string) {
    console.log(`${message}:`);
    console.log(reduceSpew(err.message ?? "No message").replace(/(^|\n)/g, "$1> "));
    console.log(reduceSpew(err.stack ?? "Unknown Stack").replace(/(^|\n)/g, "$1> "));
}

function reduceSpew(message: string): string {
    // Since this is only a warning, it tends to be reported many (i.e. thousands of) times
    const problemString = "npm WARN tar ENOSPC: no space left on device, write\n";
    const index = message.indexOf(problemString);
    if (index < 0) return message;

    return message.substring(0, index) + problemString + replaceAll(message.substring(index), problemString, "");
}

function replaceAll(message: string, oldStr: string, newStr: string) {
    let result = "";
    let index = 0;
    while (true) {
        const newIndex = message.indexOf(oldStr, index);
        if (newIndex < 0) {
            return index === 0
                ? message
                : result + message.substring(index);
        }

        result += message.substring(index, newIndex);
        result += newStr;

        index = newIndex + oldStr.length;
    }
}

function makeMarkdownLink(url: string) {
    const match = /\/blob\/[a-f0-9]+\/(.+)$/.exec(url);
    return !match
        ? url
        : `[${match[1]}](${url})`;
}

async function downloadTypeScriptAsync(cwd: string, params: GitParams | UserParams): Promise<{ oldTscPath: string, oldTscResolvedVersion: string, newTscPath: string, newTscResolvedVersion: string }> {
    if (params.testType === 'user') {
        const { tscPath: oldTscPath, resolvedVersion: oldTscResolvedVersion } = await downloadTypescriptRepoAsync(cwd, params.oldTypescriptRepoUrl, params.oldHeadRef);
        // We need to handle the ref/pull/*/merge differently as it is not a branch and cannot be pulled during clone.
        const { tscPath: newTscPath, resolvedVersion: newTscResolvedVersion } = await downloadTypescriptSourceIssueAsync(cwd, params.oldTypescriptRepoUrl, params.sourceIssue);

        return {
            oldTscPath,
            oldTscResolvedVersion,
            newTscPath,
            newTscResolvedVersion
        };
    }
    else if (params.testType === 'git') {
        const { tscPath: oldTscPath, resolvedVersion: oldTscResolvedVersion } = await downloadTypeScriptNpmAsync(cwd, params.oldTscVersion);
        const { tscPath: newTscPath, resolvedVersion: newTscResolvedVersion } = await downloadTypeScriptNpmAsync(cwd, params.newTscVersion);

        return {
            oldTscPath,
            oldTscResolvedVersion,
            newTscPath,
            newTscResolvedVersion
        };
    }
    else {
        throw new Error('Invalid parameters');
    }
}

export async function downloadTypescriptRepoAsync(cwd: string, repoUrl: string, headRef: string): Promise<{ tscPath: string, resolvedVersion: string }> {
    const repoName = `typescript-${headRef}`;
    await git.cloneRepoIfNecessary(cwd, { name: repoName, url: repoUrl, branch: headRef });

    const repoPath = path.join(cwd, repoName);

    return {
        tscPath: await buildTsc(repoPath),
        resolvedVersion: headRef
    };
}

async function downloadTypescriptSourceIssueAsync(cwd: string, repoUrl: string, sourceIssue: number): Promise<{ tscPath: string, resolvedVersion: string }> {
    const repoName = `typescript-${sourceIssue}`;
    await git.cloneRepoIfNecessary(cwd, { name: repoName, url: repoUrl });

    const repoPath = path.join(cwd, repoName);
    const headRef = `refs/pull/${sourceIssue}/merge`;

    await git.checkout(repoPath, headRef);

    return {
        tscPath: await buildTsc(repoPath),
        resolvedVersion: headRef
    };
}

async function buildTsc(repoPath: string) {
    await execAsync(repoPath, "npm ci");
    await execAsync(repoPath, "npm run build:compiler");

    // We build the LKG for the benefit of scenarios that want to install it as an npm package
    await execAsync(repoPath, "npx gulp configure-insiders");
    await execAsync(repoPath, "npx gulp LKG");

    return path.join(repoPath, "built", "local", "tsc.js");
}

async function downloadTypeScriptNpmAsync(cwd: string, version: string): Promise<{ tscPath: string, resolvedVersion: string }> {
    const tarName = (await execAsync(cwd, `npm pack typescript@${version} --quiet`)).trim();

    const tarMatch = /^(typescript-(.+))\..+$/.exec(tarName);
    if (!tarMatch) {
        throw new Error("Unexpected tarball name format: " + tarName);
    }

    const resolvedVersion = tarMatch[2];
    const dirName = tarMatch[1];
    const dirPath = path.join(processCwd, dirName);

    await execAsync(cwd, `tar xf ${tarName} && rm ${tarName}`);
    await fs.promises.rename(path.join(processCwd, "package"), dirPath);

    const tscPath = path.join(dirPath, "lib", "tsc.js");
    if (!await pu.exists(tscPath)) {
        throw new Error("Cannot find file " + tscPath);
    }

    return { tscPath, resolvedVersion };
}
