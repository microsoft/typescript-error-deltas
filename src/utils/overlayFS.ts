import fs from "fs";
import path from "path";
import { execAsync } from "./execUtils";

export interface OverlayBaseFS {
    path: string;
    createOverlay(): Promise<OverlayMergedFS>;
}

export interface OverlayMergedFS extends AsyncDisposable {
    path: string;
}

export interface DisposableOverlayBaseFS extends OverlayBaseFS, AsyncDisposable {}

const processCwd = process.cwd();

/**
 * Creates an overlay FS using a tmpfs mount. A base directory is created on the tmpfs.
 * New overlays are created by mounting an overlay on top of the base directory.
 * 
 * This requires root access.
 */
export async function createTempOverlayFS(root: string, diagnosticOutput: boolean): Promise<DisposableOverlayBaseFS> {
    await tryUnmount(root);
    await rmWithRetryAsRoot(root);
    await mkdirAllAsRoot(root);
    await execAsync(processCwd, `sudo mount -t tmpfs tmpfs ${root}`);

    const lowerDir = path.join(root, "base");
    await mkdirAll(lowerDir);

    let overlay: OverlayMergedFS | undefined;

    async function createOverlay(): Promise<OverlayMergedFS> {
        if (overlay) {
            throw new Error("Overlay has already been created");
        }

        // Using short names here as these paths can appear in the summaries.
        const overlayRoot = path.join(root, "_");
        await rmWithRetryAsRoot(overlayRoot);

        const upperDir = path.join(overlayRoot, ".u");
        const workDir = path.join(overlayRoot, ".w");
        const merged = path.join(overlayRoot, "m");
        
        await mkdirAll(upperDir, workDir, merged);

        if (diagnosticOutput) {
            await diskUsageRoot(lowerDir);
            await diskUsageRoot(overlayRoot);
        }

        await execAsync(processCwd, `sudo mount -t overlay overlay -o lowerdir=${lowerDir},upperdir=${upperDir},workdir=${workDir} ${merged}`);

        overlay = {
            path: merged,
            [Symbol.asyncDispose]: async () => {
                overlay = undefined;
                if (diagnosticOutput) {
                    await diskUsageRoot(upperDir);
                }
                await tryUnmount(merged);
                await rmWithRetryAsRoot(overlayRoot);
            }
        }

        return overlay;
    }

    return {
        path: lowerDir,
        createOverlay,
        [Symbol.asyncDispose]: async () => {
            if (diagnosticOutput) {
                await diskUsageRoot(root);
            }
            if (overlay) {
                await overlay[Symbol.asyncDispose]();
            }
            await tryUnmount(root);
            await rmWithRetryAsRoot(root);
        },  
    }
}

async function retry(fn: (() => void) | (() => Promise<void>), retries: number, delayMs: number): Promise<void> {
    for (let i = 0; i < retries; i++) {
        try {
            await fn();
            return;
        } catch (e) {
            if (i === retries - 1) {
                throw e;
            }
            await new Promise(resolve => setTimeout(resolve, delayMs));
        }
    }
}

async function tryUnmount(p: string) {
    if (!fs.existsSync(p)) return;
    try {
        await execAsync(processCwd, `sudo umount -R ${p}`)
    } catch {
        // Print out handles for debugging.
        await execAsync(processCwd, `sudo lsof ${p}`)
    }
}

function diskUsageRoot(p: string) {
    return execAsync(processCwd, `sudo du -sh ${p}`);
}

function rmWithRetry(p: string) {
    return retry(() => execAsync(processCwd, `rm -rf ${p}`), 3, 1000);
}

function rmWithRetryAsRoot(p: string) {
    return retry(() => execAsync(processCwd, `sudo rm -rf ${p}`), 3, 1000);
}

function mkdirAll(...args: string[]) {
    return execAsync(processCwd, `mkdir -p ${args.join(" ")}`);
}

function mkdirAllAsRoot(...args: string[]) {
    return execAsync(processCwd, `sudo mkdir -p ${args.join(" ")}`);
}

/**
 * Creates a fake overlay FS, which is just a directory on the local filesystem.
 * Overlays are created by copying the contents of the `base` directory.
 */
export async function createCopyingOverlayFS(root: string, _diagnosticOutput: boolean): Promise<DisposableOverlayBaseFS> {
    await rmWithRetry(root);
    await mkdirAll(root);

    const basePath = path.join(root, "base");
    await mkdirAll(basePath);

    let overlay: OverlayMergedFS | undefined;

    async function createOverlay(): Promise<OverlayMergedFS> {
        if (overlay) {
            throw new Error("Overlay has already been created");
        }

        const overlayRoot = path.join(root, "overlay");
        await rmWithRetry(overlayRoot);

        await execAsync(processCwd, `cp -r --reflink=auto ${basePath} ${overlayRoot}`);

        overlay = {
            path: overlayRoot,
            [Symbol.asyncDispose]: async () => {
                overlay = undefined;
                await rmWithRetry(overlayRoot);
            }
        }

        return overlay;
    }

    return {
        path: basePath,
        createOverlay,
        [Symbol.asyncDispose]: async () => {
            if (overlay) {
                await overlay[Symbol.asyncDispose]();
                overlay = undefined;
            }
            await rmWithRetry(root);
        },  
    }
}
