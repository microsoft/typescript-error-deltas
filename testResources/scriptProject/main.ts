export function noDefaultsAllowed(p = 1) {
    return p
}
