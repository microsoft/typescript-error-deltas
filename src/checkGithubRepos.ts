import path = require("path");
import { mainAsync, reportError, TsEntrypoint } from "./main";

const { argv } = process;

if (argv.length !== 11) {
    console.error(`Usage: ${path.basename(argv[0])} ${path.basename(argv[1])} <ts_entrypoint> <old_ts_npm_version> <new_ts_npm_version> <repo_list_path> <worker_count> <worker_number> <result_dir_name> <diagnostic_output> <prng_seed>`);
    process.exit(-1);
}

const [,, entrypoint, oldTsNpmVersion, newTsNpmVersion, repoListPath, workerCount, workerNumber, resultDirName, diagnosticOutput, prngSeed] = argv;

mainAsync({
    testType: "github",
    tmpfs: false,
    entrypoint: entrypoint as TsEntrypoint,
    diagnosticOutput: diagnosticOutput.toLowerCase() === "true",
    buildWithNewWhenOldFails: false,
    repoListPath,
    workerCount: +workerCount,
    workerNumber: +workerNumber,
    oldTsNpmVersion,
    newTsNpmVersion,
    resultDirName,
    prngSeed: prngSeed.toLowerCase() === "n/a" ? undefined : prngSeed,
}).catch(err => {
    reportError(err, "Unhandled exception");
    process.exit(1);
});
