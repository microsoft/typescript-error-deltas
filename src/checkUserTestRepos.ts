import path = require("path");
import { mainAsync, reportError, TsEntrypoint } from "./main";

const { argv } = process;

if (argv.length !== 13) {
    console.error(`Usage: ${path.basename(argv[0])} ${path.basename(argv[1])} <ts_entrypoint> <old_ts_repo_url> <old_head_ref> <pr_number> <is_top_repos> <repo_list_path> <worker_count> <worker_number> <result_dir_name> <diagnostic_output> <prng_seed>`);
    process.exit(-1);
}

const [,, entrypoint, oldTsRepoUrl, oldHeadRef, prNumber, buildWithNewWhenOldFails, repoListPath, workerCount, workerNumber, resultDirName, diagnosticOutput, prngSeed] = argv;

mainAsync({
    testType: "user",
    tmpfs: false,
    entrypoint: entrypoint as TsEntrypoint,
    oldTsRepoUrl,
    oldHeadRef,
    prNumber: +prNumber,
    buildWithNewWhenOldFails: buildWithNewWhenOldFails.toLowerCase() !== "true",
    repoListPath,
    workerCount: +workerCount,
    workerNumber: +workerNumber,
    resultDirName,
    diagnosticOutput: diagnosticOutput.toLowerCase() === "true",
    prngSeed: prngSeed.toLowerCase() === "n/a" ? undefined : prngSeed,
}).catch(err => {
    reportError(err, "Unhandled exception");
    process.exit(1);
});
