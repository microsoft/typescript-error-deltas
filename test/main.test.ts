/// <reference types="jest" />
import { getTscRepoResult, downloadTsRepoAsync } from '../src/main'
import { execSync } from "child_process"
import { existsSync, mkdirSync } from "fs"
import path = require("path")
import { createCopyingOverlayFS } from '../src/utils/overlayFS'
describe("main", () => {
    jest.setTimeout(10 * 60 * 1000)
    xit("build-only correctly caches", async () => {
        const { status, summary } = await getTscRepoResult(
            {
                name: "TypeScript-Node-Starter",
                url: "https://github.com/Microsoft/TypeScript-Node-Starter.git"
            },
            "./userTests",
            path.resolve("./typescript-main/built/local/tsc.js"),
            path.resolve("./typescript-44585/built/local/tsc.js"),
            /*ignoreOldTscFailures*/ true, // as in a user test
            await createCopyingOverlayFS("./ts_downloads", false),
            /*diagnosticOutput*/ false)
        expect(status).toEqual("NewBuildHadErrors")
        expect(summary).toBeDefined()
        expect(summary!.startsWith(`# [TypeScript-Node-Starter](https://github.com/Microsoft/TypeScript-Node-Starter.git)`)).toBeTruthy()
        expect(summary!.includes("- \`error TS2496: The 'arguments' object cannot be referenced in an arrow function in ES3 and ES5. Consider using a standard function expression.\`")).toBeTruthy()
    })
    it("downloads from a branch", async () => {
        if (!existsSync("./testDownloads/main")) {
            mkdirSync("./testDownloads/main", { recursive: true });
        }
        else if (existsSync("./testDownloads/main/typescript-test-fake-error")) {
            execSync("cd ./testDownloads/main/typescript-test-fake-error && git restore . && cd ..")
        }
        await downloadTsRepoAsync('./testDownloads/main', 'https://github.com/sandersn/typescript', 'test-fake-error', 'tsc')
    })
})
