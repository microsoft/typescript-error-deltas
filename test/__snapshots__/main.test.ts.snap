// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main outputs server errors 1`] = `
[MockFunction] {
  "calls": [
    [
      "<BASE_PATH>/ts_downloads/base/MockRepoOwner.MockRepoName.rawError.txt",
      "Req #123 - cursedCommand
Some error. Could not do something. 
Maybe a Debug fail.
    at a (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:1:1)
    at b (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:2:2)
    at c (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:3:3)
    at d (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:4:4)
    at e (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:5:5)",
    ],
    [
      "<BASE_PATH>/RepoResults123/718e48b799650d4b66e5d80ad6bac7b9.results.txt",
      "<h2>Maybe a Debug fail.</h2>

\`\`\`
Req #123 - cursedCommand
    at a (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:1:1)
    at b (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:2:2)
    at c (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:3:3)
    at d (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:4:4)
    at e (/mnt/vss/_work/1/s/typescript-1.1.1/lib/typescript.js:5:5)
\`\`\`

<h4>Affected repos</h4>
<details>
<summary><a href="https://github.com/MockRepoOwner/MockRepoName">MockRepoOwner/MockRepoName</a></summary>
Raw error text: <code>RepoResults123/MockRepoOwner.MockRepoName.rawError.txt</code> in the <a href="PLACEHOLDER_ARTIFACT_FOLDER">artifact folder</a>
<h4>Last few requests</h4>

\`\`\`json
{"rootDirPlaceholder":"@PROJECT_ROOT@","serverArgs":["--disableAutomaticTypingAcquisition"]}
{"seq":1,"type":"request","command":"configure","arguments":{"preferences":{"disableLineTextInReferences":true,"includePackageJsonAutoImports":"auto","includeCompletionsForImportStatements":true,"includeCompletionsWithSnippetText":true,"includeAutomaticOptionalChainCompletions":true,"includeCompletionsWithInsertText":true,"includeCompletionsWithClassMemberSnippets":true,"allowIncompleteCompletions":true,"includeCompletionsForModuleExports":false},"watchOptions":{"excludeDirectories":["**/node_modules"]}}}
{"seq":2,"type":"request","command":"updateOpen","arguments":{"changedFiles":[],"closedFiles":[],"openFiles":[{"file":"@PROJECT_ROOT@/sample_repoName.config.js","projectRootPath":"@PROJECT_ROOT@"}]}}
{"seq":3,"type":"request","command":"cursedCommand","arguments":{"file":"@PROJECT_ROOT@/src/sampleTsFile.ts","line":1,"offset":1,"includeExternalModuleExports":false,"triggerKind":1}}
\`\`\`

<h4>Repro steps</h4>

\`\`\`bash
#!/bin/bash

git clone https://github.com/MockRepoOwner/MockRepoName --recurse-submodules
git -C "./MockRepoName" reset --hard 57b462387e88aa7e363af0daf867a5dc1e83a935
# Install packages (exact steps are below, but it might be easier to follow the repo readme)
npm --prefix "./dirA" install --prefer-offline --no-audit --no-progress --legacy-peer-deps --ignore-scripts -q
npm --prefix "./dirB/dirC" install --prefer-offline --no-audit --no-progress --legacy-peer-deps --ignore-scripts -q
npm --prefix "./dirD/dirE/dirF" install --prefer-offline --no-audit --no-progress --legacy-peer-deps --ignore-scripts -q
downloadUrl=$(curl -s "PLACEHOLDER_GETARTIFACTS_API?artifactName=RepoResults123&api-version=7.0" | jq -r ".resource.downloadUrl")
wget -O RepoResults123.zip "$downloadUrl"
unzip -p RepoResults123.zip RepoResults123/MockRepoOwner.MockRepoName.replay.txt > MockRepoOwner.MockRepoName.replay.txt
npm install --no-save @typescript/server-replay
\`\`\`

To run the repro:
\`\`\`bash
# \`npx tsreplay --help\` to learn about helpful switches for debugging, logging, etc.
npx tsreplay ./MockRepoName ./MockRepoOwner.MockRepoName.replay.txt <PATH_TO_tsserver.js>
\`\`\`

</details>
",
      {
        "encoding": "utf-8",
      },
    ],
    [
      "<BASE_PATH>/RepoResults123/metadata.json",
      "{"newTsResolvedVersion":"1.1.1","oldTsResolvedVersion":"0.0.0","statusCounts":{"Detected interesting changes":1}}",
      {
        "encoding": "utf-8",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
